### PR TITLE
Support a "retry now" method

### PR DESCRIPTION
**Pull request by CharlieHess**
* [x] I've read and understood the [Contributing guidelines](./CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](./CODE_OF_CONDUCT.md).
* [x] I've been mindful about doing atomic commits, adding documentation to my changes, not refactoring too much.
* [x] I've a descriptive title and added any useful information for the reviewer. Where appropriate, I've attached a screenshot and/or screencast (gif preferrably).
* [x] I've written tests to cover the new code and functionality included in this PR.
* [x] I've read, agree to, and signed the [Contributor License Agreement (CLA)](https://docs.google.com/a/slack-corp.com/forms/d/1q_w8rlJG_x_xJOoSUMNl7R35rkpA7N6pUkKhfHHMD9c/viewform).

## 🚧 Note 
Merge #349 first then rebase.

## PR Summary
This PR builds on the work in #349. Since we're using a "retry forever with exponential backoff" policy, we can hit the cap of 30 minutes relatively quickly. If we then regain network or wake from sleep, we want to be able to recreate the socket connection immediately rather than waiting for the next retry. This PR adds a `forceRetry` parameter to `rtm.reconnect` to support this.

Then, in my client-side code, I can do something like:

``` js
// If we have network when we resume from sleep, retry immediately
async handleWakeFromSleep() {
  const online = await isOnline({ timeout: 30000 });
  if (online) {
    this.rtm.reconnect(true);
  }
}
```

## Related Issues
No issues filed.

## Test strategy
One new test was added:
```
 ✓ should support reconnecting immediately
```

### CODE REVIEW

File: lib/clients/client.js [hunk 1]
```diff
@@ -102,6 +98,22 @@ BaseAPIClient.prototype.registerDataStore = function registerDataStore(dataStore
...
 /**
+ * Empties the request queue and returns it to a clean slate.
+ */
+BaseAPIClient.prototype.clearQueue = function clearQueue() {
+  this.logger('info', 'Clearing the request queue');
+  this.requestQueue.kill();
+
+  // Initially tried to accomplish this with just `requestQueue.kill`, but it leaves the queue
+  // blocked. See https://github.com/caolan/async/issues/1422.
+  this.requestQueue = async.queue(
```
Comment (aoberoi) [inline lib/clients/client.js]: there's some code duplication here with L61-L63. but honestly, that can be cleaned up with a later pass.

File: lib/clients/transports/call-transport.js [hunk 2]
```diff
@@ -102,15 +101,15 @@ var handleHttpResponse = function handleHttpResponse(body, headers, client, apiC
   if (!jsonResponse.ok) {
     jsonError = new Error(jsonResponse.error);
     client.logger('error', 'Response not OK: ', jsonResponse.error);
-  }
```
Comment (aoberoi) [inline lib/clients/transports/call-transport.js]: what if the scopes are relevant to the error that just occurred, and the client wanted this data in the `jsonResponse` object?

**Comment by aoberoi** [comment]
@CharlieHess anything i can do to help this progress?

**Comment by CharlieHess** [comment]
@aoberoi after further consideration, we don't need this method. Just need to be sure to call `disconnect` before you `reconnect`.

**Pull request closed without merging**